src/cryptonote_core/pulse.cpp

```diff
@@ -687,319 +983,421 @@ event_loop prepare_for_round(round_context &context, service_nodes::service_node
 }
 
-event_loop wait_for_random_value_hashes(round_context &context)
+round_state send_and_wait_for_random_value(round_context &context, void *quorumnet_state, service_nodes::service_node_keys const &key)
 {
-  bool timed_out = pulse::clock::now() >= context.wait_for_block_template.end_time;
+  //
+  // NOTE: Send
+  //
+  assert(context.prepare_for_round.participant == sn_type::validator);
+  if (context.transient.random_value.send.one_time_only())
+  {
+    // Message
+    pulse::message msg     = msg_init_from_context(context);
+    msg.type               = pulse::message_type::random_value;
+    msg.random_value.value = context.transient.random_value.send.data;
+    crypto::generate_signature(msg_signature_hash(context, msg), key.pub, key.key, msg.signature);
+    handle_message(quorumnet_state, msg); // Add our own. We receive our own msg for the first time which also triggers us to relay.
+  }
+
+  //
+  // NOTE: Wait
+  //
+  handle_messages_received_early_for(context.transient.random_value.wait.stage, quorumnet_state);
+  pulse_wait_stage const &stage = context.transient.random_value.wait.stage;
 
-  int received_hashes             = 0;
-  int expected_hashes             = 0;
-  uint16_t received_hashes_bitset = 0;
-  uint16_t validator_bitset       = context.wait_for_block_template.block.pulse.validator_bitset;
-  for (size_t i = 0; i < service_nodes::PULSE_QUORUM_NUM_VALIDATORS; i++)
+  auto const &quorum    = context.transient.random_value.wait.data;
+  bool const timed_out  = pulse::clock::now() >= stage.end_time;
+  bool const all_values = stage.msgs_received == context.transient.wait_for_handshake_bitsets.best_count;
+
+  if (timed_out || all_values)
   {
-    auto &[hash, received] = context.wait_for_random_value_hashes.hashes[i];
-    uint16_t bit           = (1 << i);
+    if (!enforce_validator_participation_and_timeouts(context, stage, timed_out, all_values))
+      return goto_preparing_for_next_round(context);
 
-    if (validator_bitset & bit) // This validator is locked in for this round
+    // Generate Final Random Value
+    crypto::hash final_hash = {};
+    for (size_t index = 0; index < quorum.size(); index++)
     {
-      received_hashes_bitset |= bit;
-      received_hashes += received;
-      expected_hashes++;
+      auto &[random_value, received] = quorum[index];
+      if (received)
+      {
+        epee::wipeable_string string = lokimq::to_hex(tools::view_guts(random_value.data));
+
+#if defined(NDEBUG)
+        // Mask the random value generated incase someone is snooping logs
+        // trying to derive the Service Node rng seed.
+        for (int i = 2; i < static_cast<int>(string.size()) - 2; i++)
+          string.data()[i] = '.';
+#endif
+
+        MDEBUG(log_prefix(context) << "Final random value seeding with V[" << index << "] " << string.view());
+
+        auto buf   = tools::memcpy_le(final_hash.data, random_value.data);
+        final_hash = crypto::cn_fast_hash(buf.data(), buf.size());
```

Comment:
Rather than hashing the hash each time through, it seems like this could just build a string from concatenating all the random values* then hashing that at the end of the loop.  The hash-of-a-hash seems like some complexity that shouldn't hurt anything, but makes someone wonder why.

(Most hash function APIs can be updated piecewise, e.g. sodium's `crypto_generichash_update`, but it looks like cn_fast_hash doesn't because... yeah.  The sodium alternative is blake2b instead of SHA-512, but that might be a good choice anyway).

Reply:
Yeah good point. I took the opportunity to change all the hashing calls to sodium. So generichash key-less for some of the message passing and piecewise key-less hashing for this step.